RIPS/rip-9995.md

```diff
@@ -0,0 +1,199 @@
+
+We define the following "action" identifiers:
+
+```
+sender_deployment = 0x01
+sender_validation = 0x02
+paymaster_validation = 0x03
+paymaster_posttx = 0x04
+```
+
+#### Sender Deployment
+
+Inputs to the `Deployer` contract are not defined by the protocol and are controlled by the `deployerData` parameter.
+
+The sender deployment frame MUST result in the `sender` address becoming initialized with contract code.
+
+#### Sender Validation
+
+Inputs to the `Sender` validation section are defined by the protocol as a packed encoding of the following data:
+```
+uint256 version, uint256 action, bytes32 txHash, bytes transaction
+```
+
+The sender validation frame MUST return 32 bytes `validationData` that is interpreted as:
+
+```solidity
+abi.encodePacked(MAGIC_VALUE_SENDER, validUntil, validAfter)
+```
+
+#### Paymaster Validation
+
+Inputs to the `Paymaster` validation section are defined by the protocol as a packed encoding of the following data:
+```
+uint256 version, uint256 action, bytes32 txHash, bytes transaction
+```
+
+The frame must return a bytes array that is interpreted as:
+
+```solidity
+abi.encode(context, MAGIC_VALUE_PAYMASTER, validUntil, validAfter)
+```
+
+#### Sender Execution
+As the beginning of actual transaction execution, this call is made to the `Sender` contract directly,
+meaning it will be executed starting at the first byte of code section 0, and `pc` set to 0.
+
+Inputs to the `Sender` contract are not defined by the protocol and are controlled by the `callData` parameter.
+
+#### Paymaster post-transaction frame
+
+Inputs to the `Paymaster` post-transaction are defined by the protocol as a packed encoding of the following data:
+```
+uint256 version, uint256 action, uint256 actualGasCost, bool success, bytes context
+```
+
+## Rationale
+
+### ABI encoding for system frames' input and output data
+
+Using an encoding format for data provided by the protocol itself does represent an abstraction levels violation,
+it is a relatively safe and any alternative solution would require some trade-offs.
+
+The validation section of a Smart Contract Account code needs to have full access to the majority of transaction
+details in order to be able to make an informed decision about either accepting or rejecting the transaction.
+
+A small subset of this data is available with the existing opcodes, however creating an opcode for every transaction
+parameter is not feasible.
+
+Allowing wallets to specify their own encoding for this data is also not feasible as Smart Contract Accounts must
+avoid any ambiguity about the meaning of the received data.
+
+Using a different encoding will not benefit from the fact that ABI decoding logic is already built into
+Solidity programming language and ships with most of deployed smart contracts.
+
+### Allowing regular code execution flows to access `kind_abstraction` sections
+
+The code in a `kind_abstraction` does not have any difference
```

Comment:
I think it is a feature that it is callable only from entryPoint. after all, account relies that validateUserOp is called exactly once for each execution call